src/orange/univ/SimpleUnitTracker.hh

```diff
@@ -205,29 +205,46 @@ SimpleUnitTracker::cross_boundary(LocalState const& state) const
     detail::SenseCalculator calc_senses(
         this->make_local_surfaces(), state.pos, state.temp_sense);
 
-    // Loop over all connected surfaces (TODO: intersect with BVH)
-    for (LocalVolumeId volid : this->get_neighbors(state.surface.id()))
-    {
-        if (volid == state.volume)
+    VolumeView* temp_vv;
+    detail::OnFace temp_face;
+    auto is_inside = [this, &state, &calc_senses, &temp_face, &temp_vv](
+                         LocalVolumeId const& id) -> bool {
+        if (id == state.volume)
         {
             // Cannot cross surface into the same volume
-            continue;
+            return false;
         }
-        VolumeView vol = this->make_local_volume(volid);
 
-        // Calculate the local senses and face
+        VolumeView vol = this->make_local_volume(id);
         auto logic_state
             = calc_senses(vol, detail::find_face(vol, state.surface));
 
-        // Evaluate whether the senses are "inside" the volume
-        if (!detail::LogicEvaluator(vol.logic())(logic_state.senses))
+        temp_vv = &vol;
+        temp_face = logic_state.face;
+        return detail::LogicEvaluator(vol.logic())(logic_state.senses);
```

Comment:
You're saving a pointer to a temporary here 😱

Instead, just make a local `OnLocalSurface on_surface`, and 
```suggestion
        if (detail::LogicEvaluator(vol.logic())(logic_state.senses))
        {
            // Inside: find and save the local surface ID, and end the search
            on_surface = surface(vol, logic_state.face);
            return true;
        }
        return false;
```